core/src/main/scala/com/avast/grpc/jsonbridge/ReflectionGrpcJsonBridge.scala

```diff
@@ -0,0 +1,157 @@
+import scala.util.control.NonFatal
+
+class ReflectionGrpcJsonBridge[F[_]](services: ServerServiceDefinition*)(implicit val ec: ExecutionContext, val F: Async[F])
+    extends GrpcJsonBridge[F]
+    with AutoCloseable
+    with StrictLogging {
+
+  def this(grpcServer: io.grpc.Server)(implicit ec: ExecutionContext, F: Async[F]) =
+    this(grpcServer.getImmutableServices.asScala: _*)(ec, F)
+
+  protected implicit val scheduler: Scheduler = Scheduler(ec)
+
+  protected def supportedMethod(d: ServerMethodDefinition[_, _]): Boolean = d.getMethodDescriptor.getType == MethodType.UNARY
+  protected val parser: JsonFormat.Parser = JsonFormat.parser()
+  protected val printer: JsonFormat.Printer = {
+    JsonFormat.printer().includingDefaultValueFields().omittingInsignificantWhitespace()
+  }
+
+  private val inProcessServiceName = s"HttpWrapper-${System.nanoTime()}"
+  private val inProcessServer = {
+    val b = InProcessServerBuilder.forName(inProcessServiceName).executor(scheduler)
+    services.foreach(b.addService)
+    b.build().start()
+  }
+  private val inProcessChannel = InProcessChannelBuilder.forName(inProcessServiceName).executor(scheduler).build()
+
+  override def close(): Unit = {
+    inProcessChannel.shutdownNow()
+    inProcessServer.shutdownNow()
+    ()
+  }
+
+  // JSON body and headers to a response (fail status or JSON response)
+  protected type HandlerFunc = (String, Map[String, String]) => F[Either[Status, String]]
+
+  // map from full method name to a function that invokes that method
+  protected val handlersPerMethod: Map[String, HandlerFunc] =
+    inProcessServer.getImmutableServices.asScala.flatMap { ssd =>
+      val newFutureStub = createNewFutureStubFunction(ssd)
+      val methods = ssd.getMethods.asScala
+        .filter(supportedMethod)
+        .map { m =>
+          val requestMarshaller = m.getMethodDescriptor.getRequestMarshaller.asInstanceOf[PrototypeMarshaller[_]]
+          val requestMessagePrototype = requestMarshaller.getMessagePrototype.asInstanceOf[Message]
+          val Seq(_, methodName) = m.getMethodDescriptor.getFullMethodName.split('/').toSeq
+          val javaMethodName = methodName.substring(0, 1).toLowerCase + methodName.substring(1)
+          val stubMethod = newFutureStub().getClass.getDeclaredMethod(javaMethodName, requestMessagePrototype.getClass)
+          val handler: HandlerFunc = (json, headers) =>
+            Task
+              .deferFuture {
+                val md = new Metadata()
+                headers.foreach { case (k, v) => md.put(Metadata.Key.of(k, Metadata.ASCII_STRING_MARSHALLER), v) }
+                val stubWithHeaders: AbstractStub[_] = JavaGenericHelper.attachHeaders(newFutureStub(), md)
+                val request = parseRequest(json, requestMessagePrototype)
+                request match {
+                  case Right(r) =>
+                    val listenableFuture = stubMethod.invoke(stubWithHeaders, r).asInstanceOf[ListenableFuture[MessageOrBuilder]]
```

Comment:
I've rewritten this piece of code to be more readable. See https://github.com/avast/grpc-json-bridge/commit/68f52d081a7e6b6de71f131f936ddead756379db

Reply:
We removed the dependency on Monix (thx to @sideeffffect ❤️) at the same time (-> lot of changes), so I've resolved the conflict _using mine_ and the applied some of the improvements.

Reply:
I've rewritten it again :smiley: